File: src/com/cattailsw/nanidroid/util/NetworkUtil.java
package com.cattailsw.nanidroid.util;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.zip.GZIPInputStream;

import javax.net.ssl.HttpsURLConnection;

/** HTTPS-only network boundary for archive and update downloads. */
public final class NetworkUtil {
    private static final int TIMEOUT_MILLIS = 20 * 1000;

    private NetworkUtil() {
    }

    public static boolean exists(Context context, String url) {
        HttpsURLConnection connection = null;
        try {
            connection = open(context, url);
            return connection.getResponseCode() == HttpsURLConnection.HTTP_OK;
        } catch (IOException ignored) {
            return false;
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }

    /**
     * The caller owns and must close the returned stream.  Closing it releases
     * the associated connection on Android's HTTPS implementation.
     */
    public static InputStream getURLStream(Context context, String url) throws IOException {
        HttpsURLConnection connection = open(context, url);
        int responseCode = connection.getResponseCode();
        if (responseCode < 200 || responseCode >= 300) {
            connection.disconnect();
            throw new IOException("HTTPS request failed: " + responseCode);
        }
        InputStream stream = connection.getInputStream();
        String encoding = connection.getContentEncoding();
        return "gzip".equalsIgnoreCase(encoding) ? new GZIPInputStream(stream) : stream;
    }

    private static HttpsURLConnection open(Context context, String value) throws IOException {
        URL url = requireHttps(value);
        HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
        connection.setConnectTimeout(TIMEOUT_MILLIS);
        connection.setReadTimeout(TIMEOUT_MILLIS);
        connection.setRequestProperty("Accept-Encoding", "gzip");
        connection.setRequestProperty("User-Agent", buildUserAgent(context));
        connection.setInstanceFollowRedirects(false);
        return connection;
    }

    static URL requireHttps(String value) throws IOException {
        URL url = new URL(value);
        if (!"https".equalsIgnoreCase(url.getProtocol()) || url.getHost().length() == 0) {
            throw new IOException("Only HTTPS URLs are supported");
        }
        return url;
    }

    private static String buildUserAgent(Context context) {
        if (context == null) {
            return "Nanidroid (gzip)";
        }
        try {
            PackageManager manager = context.getPackageManager();
            PackageInfo info = manager.getPackageInfo(context.getPackageName(), 0);
            return info.packageName + "/" + info.versionName + " (" + info.versionCode + ") (gzip)";
        } catch (PackageManager.NameNotFoundException ignored) {
            return "Nanidroid (gzip)";
        }
    }
}

File: src/main/kotlin/com/cattailsw/nanidroid/util/NetworkUtil.kt
package com.cattailsw.nanidroid.util

import android.content.Context
import android.content.pm.PackageManager
import java.io.IOException
import java.io.FilterInputStream
import java.io.InputStream
import java.net.URL
import java.util.zip.GZIPInputStream
import javax.net.ssl.HttpsURLConnection

class HttpStatusException(val statusCode: Int) : IOException("HTTPS request failed: $statusCode")

/** HTTPS-only network boundary for archive and update downloads. */
object NetworkUtil {
    private const val TIMEOUT_MILLIS = 20_000

    @JvmStatic
    fun exists(context: Context?, url: String?): Boolean {
        var connection: HttpsURLConnection? = null
        return try {
            connection = open(context, url)
            connection.responseCode == HttpsURLConnection.HTTP_OK
        } catch (_: IOException) {
            false
        } finally {
            connection?.disconnect()
        }
    }

    /** The caller owns and must close the returned stream. */
    @JvmStatic
    @Throws(IOException::class)
    fun getURLStream(context: Context?, url: String?): InputStream {
        val connection = open(context, url)
        val responseCode = connection.responseCode
        if (responseCode !in 200..299) {
            connection.disconnect()
            throw HttpStatusException(responseCode)
        }
        return responseStream(connection)
    }

    @Throws(IOException::class)
    internal fun responseStream(connection: HttpsURLConnection): InputStream {
        val input = try {
            connection.inputStream
        } catch (e: Exception) {
            connection.disconnect()
            throw e
        }
        val owned = object : FilterInputStream(input) {
            override fun close() {
                try {
                    super.close()
                } finally {
                    connection.disconnect()
                }
            }
        }
        return try {
            if (connection.contentEncoding.equals("gzip", ignoreCase = true)) {
                GZIPInputStream(owned)
            } else {
                owned
            }
        } catch (e: Exception) {
            try {
                owned.close()
            } catch (_: Exception) {
                connection.disconnect()
            }
            throw e
        }
    }

    @Throws(IOException::class)
    internal fun requireHttps(value: String): URL {
        val url = URL(value)
        if (!url.protocol.equals("https", ignoreCase = true) || url.host.isEmpty()) {
            throw IOException("Only HTTPS URLs are supported")
        }
        return url
    }

    @Throws(IOException::class)
    private fun open(context: Context?, value: String?): HttpsURLConnection =
        (requireHttps(value ?: throw IOException("Only HTTPS URLs are supported"))
            .openConnection() as HttpsURLConnection).apply {
            connectTimeout = TIMEOUT_MILLIS
            readTimeout = TIMEOUT_MILLIS
            setRequestProperty("Accept-Encoding", "gzip")
            setRequestProperty("User-Agent", buildUserAgent(context))
            instanceFollowRedirects = false
        }

    private fun buildUserAgent(context: Context?): String {
        if (context == null) return "Nanidroid (gzip)"
        return try {
            val info = context.packageManager.getPackageInfo(context.packageName, 0)
            "${info.packageName}/${info.versionName} (${info.versionCode}) (gzip)"
        } catch (_: PackageManager.NameNotFoundException) {
            "Nanidroid (gzip)"
        }
    }
}
